File: pptx-service/renderers/table.py
import logging

from builder import add_bg, add_textbox, add_rect, add_line, contrast_color
from theme import ThemeColors

logger = logging.getLogger(__name__)


def render(slide, data: dict, t: ThemeColors) -> None:
    add_bg(slide, t.bg)
    heading = data.get("heading", "")
    columns = data.get("columns", [])
    rows = data.get("rows", [])
    highlight_col = data.get("highlightColumn", "")
    source = data.get("source", "")

    add_textbox(slide, 0.83, 0.3, 11.0, 0.9, heading,
                t.font_display, 40, t.text, bold=True)

    if not columns:
        return

    table_x = 0.67
    table_w = 12.0
    table_top = 1.3
    col_w = table_w / len(columns)
    row_h = 0.65
    header_h = 0.7
    header_text = contrast_color(t.accent)

    add_rect(slide, table_x, table_top, table_w, header_h, fill_hex=t.accent)

    for j, col in enumerate(columns):
        cx = table_x + j * col_w
        add_textbox(slide, cx + 0.1, table_top + 0.1, col_w - 0.15,
                    header_h - 0.1, col.get("label", ""),
                    t.font_body, 20, header_text,
                    bold=True, align=col.get("align", "left"))

    if len(rows) > 6:
        logger.warning("table: %d rows provided, displaying first 6", len(rows))
    rows = rows[:6]
    for i, row in enumerate(rows):
        y = table_top + header_h + i * row_h
        is_emphasis = row.get("emphasis", False)
        row_fill = t.paper if i % 2 == 0 else t.bg
        add_rect(slide, table_x, y, table_w, row_h,
                 fill_hex=row_fill, line_hex=t.border, line_width_pt=0.25)

        cells = row.get("cells", {})
        for j, col in enumerate(columns):
            cx = table_x + j * col_w
            col_id = col.get("id", "")
            is_hl = col_id == highlight_col
            color = t.accent if is_hl else t.text
            add_textbox(slide, cx + 0.1, y + 0.1, col_w - 0.15,
                        row_h - 0.1, cells.get(col_id, ""),
                        t.font_body, 20, color,
                        bold=is_hl or is_emphasis,
                        align=col.get("align", "left"))

    if source:
        add_textbox(slide, 0.83, 7.1, 10.0, 0.35, f"Source: {source}",
                    t.font_body, 14, t.muted)

File: lib/pptx/slides/table.ts
import type { TableSlide } from '@/lib/types';
import {
  px, py, SLIDE_W,
  setBackground, addRect, addLine, addTextBox, addHeading, addPageNumber,
  type PptxTheme,
} from '../helpers';
import type PptxGenJS from 'pptxgenjs';
type PSlide = ReturnType<InstanceType<typeof PptxGenJS>['addSlide']>;

export function render(slide: PSlide, data: TableSlide, t: PptxTheme, n: number, total: number): void {
  setBackground(slide, '#' + t.bg);
  addHeading(slide, data.heading, t, py(60));

  const cols = data.columns.slice(0, 5);
  const rows = data.rows.slice(0, 6);
  const tableLeft = t.paddingIn;
  const tableW = SLIDE_W - 2 * t.paddingIn;
  const colW = tableW / cols.length;
  const headerH = py(70);
  const rowH = py(680) / Math.max(rows.length, 1);
  const startY = py(190);

  cols.forEach((col, ci) => {
    const x = tableLeft + ci * colW;
    addRect(slide, x, startY, colW - px(2), headerH, '#' + t.accent);
    addTextBox(slide, col.label, {
      x: x + px(12), y: startY + py(14), w: colW - px(24),
      fontSize: t.bodySize, color: 'FFFFFF', fontFace: t.fontBody, bold: true, align: col.align ?? 'left',
    });
  });

  rows.forEach((row, ri) => {
    const y = startY + headerH + ri * rowH;
    const isEven = ri % 2 === 1;
    if (isEven || row.emphasis) {
      addRect(slide, tableLeft, y, tableW, rowH, '#' + (row.emphasis ? t.accent : t.rule), { transparency: row.emphasis ? 85 : 92 });
    }
    cols.forEach((col, ci) => {
      const x = tableLeft + ci * colW;
      const isHL = data.highlightColumn === col.id;
      if (isHL) addLine(slide, x, y, x, y + rowH, '#' + t.accent, 3);
      addTextBox(slide, row.cells[col.id] ?? '', {
        x: x + px(12) + (isHL ? px(3) : 0), y: y + py(14), w: colW - px(24),
        fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, align: col.align ?? 'left',
      });
    });
  });

  if (data.source) {
    addTextBox(slide, data.source, {
      x: tableLeft, y: py(920), w: tableW,
      fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody,
    });
  }

  addPageNumber(slide, n, total, t);
}
